modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/reducer/MergeSortDistributedCacheQueryReducer.java

```diff
@@ -0,0 +1,107 @@
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.cache.query.reducer;
+
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.NoSuchElementException;
+import java.util.PriorityQueue;
+import java.util.Set;
+import java.util.UUID;
+import org.apache.ignite.IgniteCheckedException;
+import org.apache.ignite.cluster.ClusterNode;
+import org.apache.ignite.internal.processors.cache.query.CacheQueryPageRequester;
+import org.apache.ignite.internal.processors.cache.query.GridCacheQueryFutureAdapter;
+
+/**
+ * Reducer of distirbuted query that sort result through all nodes. Note that it's assumed that every node
+ * returns pre-sorted collection of data.
+ */
+public class MergeSortDistributedCacheQueryReducer<R> extends AbstractDistributedCacheQueryReducer<R> {
+    /**
+     * Queue of pages from all nodes. Order of streams is set with {@link #pageCmp}.
+     */
+    private final PriorityQueue<NodePage<R>> nodePages;
+
+    /**
+     * Compares head pages from all nodes to get the lowest value at the moment.
+     */
+    private final Comparator<NodePage<R>> pageCmp;
+
+    /**
+     * @param rowCmp Comparator to sort query results from different nodes.
+     */
+    public MergeSortDistributedCacheQueryReducer(
+        GridCacheQueryFutureAdapter fut, long reqId, CacheQueryPageRequester pageRequester,
+        Object queueLock, Collection<ClusterNode> nodes, Comparator<R> rowCmp
+    ) {
+        super(fut, reqId, pageRequester, queueLock, nodes);
+
+        pageCmp = (o1, o2) -> rowCmp.compare(o1.head(), o2.head());
+
+        nodePages = new PriorityQueue<>(nodes.size(), pageCmp);
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean hasNext() throws IgniteCheckedException {
+        // Initial sort.
+        if (nodePages.isEmpty() && !streams.isEmpty()) {
+            Set<UUID> nodes = new HashSet<>(streams.keySet());
+
+            for (UUID nodeId : nodes)
+                fillWithPage(nodeId);
+        }
+
+        return !nodePages.isEmpty();
+    }
+
+    /** {@inheritDoc} */
+    @Override public R next() throws IgniteCheckedException {
+        if (nodePages.isEmpty())
+            throw new NoSuchElementException("No next element. Please, be sure to invoke hasNext() before next().");
+
+        NodePage<R> page = nodePages.poll();
+
+        R o = page.next();
+
+        if (page.hasNext())
+            nodePages.offer(page);
+        else
+            fillWithPage(page.nodeId());
+
+        return o;
+    }
+
+    /** */
+    private void fillWithPage(UUID nodeId) throws IgniteCheckedException {
+        NodePage<R> page = streams.get(nodeId).nextPage();
+
+        if (!page.hasNext())
+            streams.remove(nodeId);
+        else
+            nodePages.offer(page);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void cancel() {
+        super.cancel();
+
+        nodePages.clear();
```

Comment:
Not in sync with the fillWithPage() method.

Reply:
Seems, it is possible to lost last pages silently on concurrent cancellation.

Reply:
Hi, this is OK. If a query is cancelled it means that cache query future is already done and we don't need more pages. Cancellation is invoked whether query is finished with timeout, if user closes cursor, if iterator reaches user specified limit, or with any exception that stops this future. So no need to wait those pages.